cores/esp8266/heap.cpp

```diff
@@ -23,370 +126,540 @@ extern "C" void z2EapFree(void *ptr, const char* file, int line) __attribute__((
+// !defined(DEBUG_ESP_OOM))", we only capture details for LIBC calls.
+struct umm_last_fail_alloc {
+    const void *addr;
+    size_t size;
+} _umm_last_fail_alloc = {NULL, 0};
+#endif
+
+
+///////////////////////////////////////////////////////////////////////////////
+// OOM - DEBUG_ESP_OOM extends monitoring for OOM to capture caller information
+// across various Heap entry points and their aliases.
+//
+// data capture wrapper macros and defines
+// Debugging helper, save the last caller address that got a NULL pointer
+// response. And when available, the file and line number.
 #if defined(DEBUG_ESP_OOM)
-#define PTR_CHECK__LOG_LAST_FAIL_FL(p, s, f, l) \
-    if(0 != (s) && 0 == p)\
-    {\
-      umm_last_fail_alloc_addr = __builtin_return_address(0);\
-      umm_last_fail_alloc_size = s;\
-      umm_last_fail_alloc_file = f;\
-      umm_last_fail_alloc_line = l;\
+
+// OOM - Debug printing
+//
+// IRQ/ISR safe printing macros. Printing is controled according to the results
+// of system_get_os_print(). Also, being in an IRQ may prevent the printing of
+// file names stored in PROGMEM. The PROGMEM address to the string is printed in
+// its place.
+#define DEBUG_HEAP_PRINTF ets_uart_printf
+inline bool withinISR(uint32_t ps) {
+     return ((ps & 0x0f) != 0);
+}
+
+static void IRAM_ATTR print_loc(bool inISR, size_t size, const char* file, int line, const void* caller) {
+    if (system_get_os_print()) {
+        DEBUG_HEAP_PRINTF(":oom %p(%d), File: ", caller, (int)size);
+        if (file) {
+            // Small code reduction by assuming file address is in PROGMEM
+            if (inISR) {
+                DEBUG_HEAP_PRINTF("%p", file);
+            } else {
+                size_t sz = strlen_P(file);
+                char buf[sz + 1];
+                strcpy_P(buf, file);
+                DEBUG_HEAP_PRINTF(buf);
+            }
+        } else {
+            DEBUG_HEAP_PRINTF("??");
+        }
+        DEBUG_HEAP_PRINTF(":%d\n", line);
     }
-#define PTR_CHECK__LOG_LAST_FAIL(p, s) \
-    if(0 != (s) && 0 == p)\
-    {\
-      umm_last_fail_alloc_addr = __builtin_return_address(0);\
-      umm_last_fail_alloc_size = s;\
-      umm_last_fail_alloc_file = NULL;\
-      umm_last_fail_alloc_line = 0;\
+}
+
+static bool IRAM_ATTR oom_check__log_last_fail_atomic_psflc(void *ptr, size_t size, const char* file, int line, const void* caller) {
+    if (0 != (size) && 0 == ptr) {
+        // Ensure changes to umm_last_fail_alloc are atomic. Otherwise, with a
+        // poorly timed interrupt, OOM details would mix between foreground and
+        // IRQ paths.
+        uint32_t saved_ps = xt_rsil(DEFAULT_CRITICAL_SECTION_INTLEVEL);
+        _umm_last_fail_alloc.addr = caller;
+        _umm_last_fail_alloc.size = size;
+        _umm_last_fail_alloc.file = file;
+        _umm_last_fail_alloc.line = line;
+        print_loc(withinISR(saved_ps), size, file, line, caller);
+        xt_wsr_ps(saved_ps);
```

Comment:
Just a refresher, I may be forgetting order of operation rules here...
Since we are between func calls, no need to explicitly put mem barrier to *force* variable assignment to happen after xt\_rsil and before xt\_wsr\_ps?

Reply:
Yes, that is my understanding as well. 
Also, both macros `xt_rsil` and` xt_wst_ps` include a memory barrier, `... : "memory")`, so I think we are good.